### PR TITLE
fix : #1228 Invalid character

### LINKED ISSUE #1228: [ERR_INVALID_CHAR]: Invalid character in header content ["location"]
Initialization only ：<https://codesandbox.io/s/lingering-worker-bn37n>

Resulting in the wrong url : <https://bn37n-3000.sse.codesandbox.io/中文/>
 
Related Code :
https://github.com/sveltejs/kit/blob/2e1b8c6fe8b96f56751b88d4374055d359788e03/packages/kit/src/runtime/server/index.js#L18

image

```log
TypeError [ERR_INVALID_CHAR]: Invalid character in header content ["location"]
    at ServerResponse.setHeader (_http_outgoing.js:561:3)
    at ServerResponse.writeHead (_http_server.js:289:21)
    at Immediate.<anonymous> (file:///sandbox/node_modules/@sveltejs/kit/dist/chunks/index.js:3429:11)
```

### PR DESCRIPTION
**Pull request by 2234839**
Fixes #1228 

### Before submitting the PR, please make sure you do the following
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts

### CODE REVIEW

File: packages/kit/src/runtime/server/index.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ export async function respond(incoming, options, state = {}) {
 		return {
 			status: 301,
 			headers: {
-				location: incoming.path.slice(0, -1) + (q ? `?${q}` : '')
+				location: encodeURI(incoming.path.slice(0, -1)) + (q ? `?${q}` : ''))
```
Comment (benmccann) [inline packages/kit/src/runtime/server/index.js:18]: this code is invalid. you added one start paren and two closing parens

**Comment by benmccann** [comment]
https://stackoverflow.com/questions/47687379/what-characters-are-allowed-in-http-header-values does suggest URI percent encoding, so this seems like a reasonable fix

**Comment by 2234839** [comment]
> 
> 
> https://stackoverflow.com/questions/47687379/what-characters-are-allowed-in-http-header-values does suggest URI percent encoding, so this seems like a reasonable fix

If you use encodeURIComponent for the path `/中文/` you will get `location:"%2F%E4%B8%AD%E6%96%87%2F"` In my attempt the browser does not handle it correctly, so `/` should not be escaped

**Comment by Rich-Harris** [comment]
Thank you — could you add a test and a changeset please? For example you could add a test to [this file](https://github.com/sveltejs/kit/blob/master/packages/kit/test/apps/basics/src/routes/routing/__tests__.js) that verifies [this page](https://github.com/sveltejs/kit/blob/master/packages/kit/test/apps/basics/src/routes/routing/%5Bslug%5D.svelte) renders as expected if you visit e.g. `/routing/中文`

**Comment by 2234839** [comment]
> 
> 
> Thank you — could you add a test and a changeset please? For example you could add a test to [this file](https://github.com/sveltejs/kit/blob/master/packages/kit/test/apps/basics/src/routes/routing/__tests__.js) that verifies [this page](https://github.com/sveltejs/kit/blob/master/packages/kit/test/apps/basics/src/routes/routing/%5Bslug%5D.svelte) renders as expected if you visit e.g. `/routing/中文`

I found a new problem when trying to add test cases,
Using a file structure like `/routing/nonascii_path/中文.svelte`, accessing `/routing/non_ascii_path/中文` 

https://github.com/sveltejs/kit/blob/52bef766133edaeae8be76d8ce689ca048bf8dc7/packages/kit/src/runtime/server/index.js#L51

The pattern here is something like `/^\/routing\/non_ascii_path\/%E4%B8%AD%E6%96%87\/?$/` , causing it to fail to match request path:`/routing/non_ascii_path/中文`, but the logic that triggers the redirect later matches request.path:`/routing/non_ascii_path/%E4%B8%AD%E6%96%87` 

I'm not sure what to do to fix this, add `decodeURI` to the match or add support for non-ascii paths where `options.manifest.routes` is generated, or something else

**Comment by benmccann** [comment]
`pnpm lint` is failing. You need to run `pnpm format`

I'm not sure what's encoding the pattern, but it sounds to me like it should be removed and that `options.manifest.routes` should support non-ascii paths

**Comment by 2234839** [comment]
> 
> 
> `pnpm lint` is failing. You need to run `pnpm format`

Okay, it's done. I thought it was a simple change and didn't download it locally for these steps

**Comment by Rich-Harris** [comment]
closing this in favour of #1273, which includes these commits but also handles other related issues (and adds tests)

**Pull request closed without merging**